src/btwa_api/app/db/firebase.py

```diff
@@ -0,0 +1,62 @@
+import json
+from collections import namedtuple
+from threading import RLock
+
+import firebase_admin
+from firebase_admin import firestore, storage
+from google.cloud.firestore_v1 import DocumentReference
+from starlette.requests import Request
+
+from ..config import FIREBASE_STORAGE_BUCKET
+
+
+class Firebase:
+    def __init__(self, bucket: str):
+        # bucket takes a very long time to initialize, so it is cached here
+        self.bucket = storage.bucket(bucket)
+        self.client = firestore.client()
+        self.users = self.client.collection("users")
+
+    def get_user_by_phone(self, phone: str) -> DocumentReference:
+        for doc in self.users.where("phone", "==", phone).stream():
+            return doc
+        return None
+
+    def get_proximity_by_phone(self, phone: str):
+        user = self.get_user_by_phone(phone)
+        if user is None:
+            return None
+        filepath = f"proximity/{user.get('uid')}.json"
+        blob = self.bucket.get_blob(filepath)
+        if blob is not None:
+            return json.loads(blob.download_as_string())
+        return None
+
+    def mark_as_infected(self, phone: str):
+        user = self.get_user_by_phone(phone)
+        if user is None:
+            return False
+        user.set({
+            "infected": True
+        })
+        return True
+
+
+FIREBASE_INSTANCE = None
```

Comment:
1) Do we absolutely need a global instance?
2) Is the single instance safe to be used across multiple threads? (given that FastAPI will run our handlers in a thread pool, because the handlers are not async)
3) Doesn't FastAPI have a better way of managing global dependencies, e.g. through its dependency injection?

Reply:
I don't like this solution, but the firebase app needs to be registered exactly once (unless I create a new one on each request) and the Storage bucket takes around 30 seconds to create, so I need to cache it. Maybe there is something wrong, it's weird that the bucket initializes so slowly. @davidvavra do you see anything wrong here?

I didn't find any information about thread safety of Firebase Admin SDK in Python sadly.